src/Compilers/CSharp/Portable/Compiler/MethodBodySynthesizer.Lowered.cs

```diff
@@ -253,6 +251,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
     /// </summary>
     internal static partial class MethodBodySynthesizer
     {
+        /// <summary>
+        /// Given a set of fields, produce an expression that is true when all of the given fields on
+        /// `this` are equal to the fields on <paramref name="otherReceiver" /> according to the
+        /// default EqualityComparer.
+        /// </summary>
+        public static BoundExpression GenerateFieldEquals<TList>(
+            BoundExpression? initialExpression,
+            BoundExpression otherReceiver,
+            TList fields,
+            SyntheticBoundNodeFactory F) where TList : IReadOnlyList<FieldSymbol>
+        {
+            Debug.Assert(fields.Count > 0);
+
+            //  Expression:
+            //
+            //      System.Collections.Generic.EqualityComparer<T_1>.Default.Equals(this.backingFld_1, value.backingFld_1)
+            //      ...
+            //      && System.Collections.Generic.EqualityComparer<T_N>.Default.Equals(this.backingFld_N, value.backingFld_N)
+
+            //  prepare symbols
+            var equalityComparer_get_Default = F.WellKnownMethod(
+                WellKnownMember.System_Collections_Generic_EqualityComparer_T__get_Default);
+            var equalityComparer_Equals = F.WellKnownMethod(
+                WellKnownMember.System_Collections_Generic_EqualityComparer_T__Equals);
+
+            NamedTypeSymbol equalityComparerType = equalityComparer_Equals.ContainingType;
+
+            BoundExpression? retExpression = initialExpression;
+
+            // Compare fields
+            foreach (var field in fields)
+            {
+                // Prepare constructed comparer
+                var constructedEqualityComparer = equalityComparerType.Construct(field.Type);
+
+                // System.Collections.Generic.EqualityComparer<T_index>.
+                //   Default.Equals(this.backingFld_index, local.backingFld_index)'
+                BoundExpression nextEquals = F.Call(
+                    F.StaticCall(constructedEqualityComparer,
+                                 equalityComparer_get_Default.AsMember(constructedEqualityComparer)),
+                    equalityComparer_Equals.AsMember(constructedEqualityComparer),
+                    F.Field(F.This(), field),
+                    F.Field(otherReceiver, field));
+
+                // Generate 'retExpression' = 'retExpression && nextEquals'
+                retExpression = retExpression is null
+                    ? nextEquals
+                    : F.LogicalAnd(retExpression, nextEquals);
+            }
+
+            RoslynDebug.AssertNotNull(retExpression);
```

Comment:
>AssertNotNull [](start = 24, length = 13)

This assertion does not appear justified.

Reply:
Earlier `fields` is asserted to have more than 1 element, there will always be at least one iteration of the `foreach`.

Reply:
How is there always at least one iteration of the foreach?  According to the spec a record can have an empty parameter list.

---
In reply to: [418268985](https://github.com/dotnet/roslyn/pull/43626#discussion_r418268985) [](ancestors = 418268985)